sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -1408,11 +1409,37 @@ case class ValidateExternalType(child: Expression, expected: DataType)
 
   override def dataType: DataType = RowEncoder.externalDataTypeForInput(expected)
 
-  override def eval(input: InternalRow): Any =
-    throw new UnsupportedOperationException("Only code-generated evaluation is supported")
-
   private val errMsg = s" is not a valid external type for schema of ${expected.simpleString}"
 
+  private lazy val checkType = expected match {
+    case _: DecimalType =>
+      (value: Any) => {
+        Seq(classOf[java.math.BigDecimal], classOf[scala.math.BigDecimal], classOf[Decimal])
+          .exists { x => value.getClass.isAssignableFrom(x) }
+      }
+    case _: ArrayType =>
+      (value: Any) => {
+        value.getClass.isAssignableFrom(classOf[Seq[_]]) || value.getClass.isArray
+      }
+    case _ if ScalaReflection.isNativeType(expected) =>
+      (value: Any) => {
+        value.getClass.isAssignableFrom(ScalaReflection.classForNativeTypeOf(expected))
+      }
+    case _ =>
+      (value: Any) => {
+        value.getClass.isAssignableFrom(dataType.asInstanceOf[ObjectType].cls)
```

Comment:
Similar as above. Let's move `dataType.asInstanceOf[ObjectType].cls` out of the function and into a `val`.

Also where is this code path in the code generated version?

Reply:
Since I thought the previous version is some difficult to read , I cleaned up the code along with the generated version. How about it?